design-proposal: improved handling of the VM FW ID (#347)
kubevirt/community: design-proposals/persist-vmi-firmware-uuid.md

# Overview
This proposal introduces a mechanism to persist the firmware UUID of a Virtual Machine in KubeVirt.
By storing the firmware UUID, we ensure that it remains consistent across VM restarts.

> which is crucial for applications and services that rely on the UUID for identification or licensing purposes.

Reviewer: While this is correct, I don't think that's the most important part of this proposal, but rather that the firmware UUID to become, well, a "real" UUID (universally unique ID). Can you please rephrase?

## Motivation
* Duplicate UUIDs cause burden with third-party automation frameworks such as Gitops tools, making it difficult to maintain unique VM UUID. 

>   Such tools have to manually assign unique UUIDs to VMs, thus also manage them externally. (issue)

Reviewer 2: I don't understand what `(issue)` means here. Can you elaborate?

  Such tools have to manually assign unique UUIDs to VMs, thus also manage them externally. (issue)
* The non-persistent nature of UUIDs may trigger redundant cloud-init setup upon VM restore. (issue)
* The non-persistent nature of UUIDs may trigger redundant license activation events during VM lifetime i.e. Windows key activation re-occur when power off and power on the VM.

## Goals
* Improve the uniqueness of the VMI firmware UUID to become independent of the VM name and namespace.
* Maintain the UUID persistence over the VM lifecycle.
* Maintain backward compatibility

> * Maintain UUID persistence with VM backup/restore.

Reviewer: Correct me if I'm wrong, but I think we won't be able to achieve this.
Perhaps this can be moved to the non-goals section.

## Non Goals

> 

Reviewer: Change UUID of currently existing VMs

## Definition of Users

> VM owners: who require consistent firmware UUIDs for their applications.

Reviewer: Also cluster-admin: to ensure VMs have universally unique firmware IDs

## User Stories
### Supported cases:

> * Creating and Starting a New VM

Reviewer 2: Would you elaborate the story? I think it is important to state that defining a VM called X in a namespace called Y in two different clusters must created two different UUIDs

> * starting and old VM that was not running

Reviewer: do we want to support that?
how would the controller distinguish between VMs that have started and VMs that have not?

Author: we would use the legacy uuid calculation for both running/stopped vms.

> * starting existing VM that was not running

Reviewer 2: Would you elaborate? What does the user expect to see?

Maybe you mean to what happens when a VM was created by an old kubevirt version and started up in a fixed version?

Author: done

## Repos
Kubevirt/kubevirt

> ## Proposed Solution

Reviewer 2: I like this solution, but what about the formerly proposed solutions? We should keep them in this proposal, and explain why this solution is better.

Author: done

### Description
The firmware UUID persistence mechanism will operate as follows:

1. **New VMs**:
    - If the firmware UUID is not explicitly defined in `vm.spec.template.spec.firmware.uuid`, the mutator webhook will automatically set the firmware UUID to the value of `vm.metadata.uid`.

> 2. **Old VMs**:

Reviewer: By "old VMs" I think you mean "existing VMs", right?

    - For VMs created before the upgrade, the VM controller will patch the `vm.spec.template.spec.firmware.uuid` field,
   with a value calculated using the legacy logic (based on the VM name).

>    This ensures backward compatibility.

Reviewer 3: Apologies if it's been brought up before, but I don't think it's great to auto-modify VM spec, especially VM template spec. Can't we just set the *VMI* firmware UUID on VMI creation to either `vm.spec.template.spec.firmware.uuid` if set or `vm.metadata.uid` if not?

Author: main issue with that approach is the backward compatibility, existing vms that didn't set `vm.spec.template.spec.firmware.uuid` will receive new FW ID on next vm restart after the upgrade

Reviewer 3: Oh of course, got it thanks!

3. **Mitigation**:
    - Backups created before the implementation of this mechanism will result in VMs receiving a new UUID upon restore, as the firmware UUID field will be absent in older backups.

>     - Users must be aware of this limitation and plan accordingly.

Reviewer 2: can we hint here what they should be doing?
It would be nice if we can have a script that adds the buggy firmware UUID on existing snapshots.

1. **Mutator Webhook**:
    - During the creation of a new VM, if the `vm.spec.template.spec.firmware.uuid` field is not set, the webhook will set it to `vm.metadata.uid`.

2. **Controller Logic**:
    - For old VMs:
        - If the firmware UUID field is absent, the controller will calculate the UUID using the legacy logic and patch the VM template spec.
        - This ensures backward compatibility without disrupting existing workflows.

3. **Backup and Restore**:
    - Backups created before this mechanism will not include the firmware UUID, leading to new UUIDs being generated upon restore.
    - Users must be aware of this limitation and plan accordingly.

--- 

## Scalability
The proposed changes have no anticipated impact on scalability capabilities of the KubeVirt framework

## Update/Rollback Compatibility
Backups created before implementing the persistent firmware UUID mechanism will not include the firmware UUID in the VM's spec.
As a result, restoring such backups will generate a new UUID for the VM.
This change may lead to compatibility issues for workloads or systems that rely on consistent UUIDs, such as licensing servers or configuration management systems.

> Users are advised to take this into consideration and plan backup and restore operations accordingly. 

Reviewer: Perhaps we can tweak and restore controller's code such that if a backup did not consist a firmware UUID in the spec, the controller would know to set up the UUID in the old way. This will ensure that old backups are compatible.

FYI @akalenyu @ShellyKa13

Author: I was able to trace the restore code to this [point:](https://github.com/kubevirt/kubevirt/blob/ce053b7e511be5cab594e65d1cc387c0c4491a96/pkg/storage/snapshot/restore.go#L685)
and it seems like this could be a potential area to manipulate the restored VM.
@akalenyu @ShellyKa13, could you please provide an estimate or insight on whether this approach is feasible? Your input would be greatly appreciated!

Author: After discussion with @mhenriks, it appears that the approach to tweak the missing uuid in snapshots is a bad practice, snapshots should be immutable once created, the snapshot can be base for restoring more then one VM which will result all those vms to have same firmware uuid (assuming those VMs are restored to different namespaces so their names wont collide).

## Testing

### Existing Tests
- Verify that a newly created VMI has a unique firmware UUID assigned.
- Ensure the UUID persists across VMI restarts.

### Scenarios Needing Coverage

1. **Old VM Patching**:  
   Validate that old VMs without a firmware UUID receive one calculated using the legacy logic.

2. **New VM Creation**:  
   Verify that the firmware UUID is set to `vm.metadata.uid` when not explicitly defined.

# Implementation Phases

1. **Alpha Phase**
    - Persist firmware UUID in the `VM` template spec field.
    - Implement logic for:
        - **New VMs**: Auto-generate and store UUID using `vm.metadata.uid` if not explicitly provided.
        - **Existing VMs**: Patch the `vm.spec.template.spec.firmware.uuid` field using the legacy calculation.
    - Cover implementation with unit tests.

2. **Beta Phase**
    - Ensure seamless upgrade compatibility for pre-existing VMs.
    - Communicate changes through release notes and detailed documentation.

3. **GA Phase**

>     - Consider deprecating the legacy logic in the future and communicate timelines accordingly.

Reviewer 2: Let us define this here explicitly: in two versions we can assume that all clusters are already upgraded and the old code can be dropped.

Author: done

# Overview
This proposal introduces a mechanism to persist the firmware UUID of a Virtual Machine Instance (VMI) in KubeVirt.
By storing the firmware UUID, we ensure that it remains consistent across VMI restarts.

> which is crucial for applications and services that rely on the UUID for identification or licensing purposes.

Reviewer 4: I wonder if it's worth mentioning cloud-init, as that would be the main issue with the current approach since the firmware UUID directly corresponds to the instance ID: https://github.com/kubevirt/kubevirt/blob/ced9fe1b5104daa431bed6b65ea690639fe3b79f/pkg/cloud-init/cloud-init.go#L117-L122

Author: not sure i understand how the cloud-init impact the scope of this proposal
could you please elaborate more on this?

Reviewer 4: Sorry if it wasn't clear. 

You wrote that the firmware UUID is crucial for applications and service that rely on it for identification or licensing purposes. I raised that cloud-init, which is perhaps the most common case of using the firmware UUID for anything, would be a good example.

Reviewer 5: Or even registration to Red Hat Satellite would be a typical application where VMs can be registered and relies on the host UUID.

## Motivation
By definition, UUID must be a universally-unique identifier.
The current implementation for the automatically-computed Firmware UUID is not universally unique: 

> Two VMs with the same name get the same UUID, even if they reside in different namespace or different clusters.

Reviewer 4: ```suggestion
Two VMs with the same name get the same UUID, even if they reside in different namespaces or different clusters.
```

Author: fixed

> This proposal discusses how to fix this, while ensuring that already-identified VMs keep their current ID.

Reviewer 4: ```suggestion
This proposal discusses how to fix this, while ensuring that existing VMs keep their current ID.
```

Author: fixed

## Goals
Generate and persist universally-unique Firmware.UUID for new VMs

> Keep the Firmware.UUID of pre-existing VMs

Author: fixed

## Non Goals

## Definition of Users
End Users: Individuals or organizations running VMs and VMIs on KubeVirt who require consistent firmware UUIDs for their applications.

## User Stories
As an end-user, I expect my VMI to maintain its identity across restarts.

## Repos
Kubevirt/kubevirt

> # Design

Reviewer 4: The introduction of a persistent, universally unique firmware UUID can be effectively divided into two key phases:

1. The initial focus on minimizing workload disruption by preserving the current generation method and persisting the UUID, followed by:
2. Transitioning to the generation of a universally unique firmware UUID.

Each phase offers multiple implementation approaches (such as using `namespace+name` for UUID generation or opting for a pseudo-random hash). To ensure clarity and thorough consideration, each phase should be addressed in separate subsections within the Design section.

There are a few possible solutions:

1. **Persist UUID in the VM’s Spec Field**  
   **Description:** The UUID would be generated upon the VM’s first boot and stored in the VM’s spec field.

   **Pros:**
   - Straightforward to implement.
   - Avoids adding a new API field, fully backward compatible.

   **Cons:**
   - This approach "abuses" the spec by storing data not representing the user's desired state.
   - It may lead to cluttered and harder-to-read VM definitions.

2. **Generate UUID via Webhook and Store in the Spec Field**  
   **Description:** If no UUID is specified by the user, a webhook generates a random UUID and assigns it to the VM’s spec field.

   **Pros:**
   - Similar to the first option (no new API field, backward compatible).

   **Cons:**
   - Adds complexity by involving webhooks, which could slow down performance if numerous VMs start simultaneously.
   - May potentially bottleneck `virt-api`.

3. **Add a Firmware UUID Field to VM Status**  
   **Description:** Store the generated UUID within the VM’s status.

   **Pros:**
   - Keeps the spec clean.
   - Focuses on VM status for generated information.

>    **Cons:**

Reviewer: > A drawback here in my view is that an object's status is often dynamic and therefore somewhat fragile. This makes sense, as controllers manipulate the status field to represent the object's current state

It is a matter of how we implement the field and the controllers that would interact with it.
I mean, as I see it, this status field will be only ever assigned once then never modified again by any controller. In this scenario I don't think it would be highly dynamic and fragile.

> while they rarely—if ever—modify the object's spec.

And for a good reason :)
Ideally, only a human should ever modify a VM's spec.

>    **Cons:**

Reviewer 6: Multiple k8s objects rely on their status fields to function. It's not that fragile and is not editable for users.

>    **Cons:**

Reviewer: > This acknowledges that disruption (i.e. firmware UUID changes post-restart to affect anything that relies on it) is imminent, correct?

This disruption has to be prevented. Otherwise it will be a breaking change.
In order to prevent it I believe we'd have to apply a multi-phased approach:
1. Continue with the current firmware UUID calculation, but store the UUID in status.
2. As @vladikr suggested, use alerts / VM conditions in order to notify the users to restart VMs that do not have firmware UUID in their status.
3. After a while, let the community know (i.e. via a mail to mailing list) that in the next version firmware UUID must reside in status in order to prevent breaking workloads.
4. Switch to using `vm.metadata.uid` as the new firmware UUID (unless a UUID is already provided in the VM's status, therefore keeping backward compatibility).

>    **Cons:**

Reviewer 4: > As @vladikr suggested, use alerts / VM conditions in order to notify the users to restart VMs that do not have firmware UUID in their status.

So at least a restart is required to derive the firmware UUID from the VMI (including shutdown VMs) so that it would persist in status, correct?

> Switch to using vm.metadata.uid as the new firmware UUID (unless a UUID is already provided in the VM's status, therefore keeping backward compatibility).

Assuming a scenario where n existing VMs in cluster(s) suffer from the issue of overlapping firmware UUIDs, what is the recommended approach for users to rectify the issue post-upgrade? Recreate the VMs? Change the `spec.firmware.uuid` to what appears in `vm.metadata.uid` and restart?

>    **Cons:**

Reviewer: > So at least a restart is required to derive the firmware UUID from the VMI (including shutdown VMs) so that it would persist in status, correct?

Now that I'm re-thinking about it, a restart might not be necessary, as virt-controller could simply copy `spec.firmware.uuid` to the status field if exists, and if not, re-calc the hash and assign it to status.

> Assuming a scenario where n existing VMs in cluster(s) suffer from the issue of overlapping firmware UUIDs, what is the recommended approach for users to rectify the issue post-upgrade? Recreate the VMs? Change the spec.firmware.uuid to what appears in vm.metadata.uid and restart? Is user action mandatory to resolve the issue?

TBH I'm not sure that we can redeem this situation. If two VMs are already running with the same firmware UUID, I think one has to be re-created. In any case, I think it shouldn't be in the scope of this proposal which only aims to prevent this situation from happening in the future.

>    **Cons:**

Reviewer 4: > TBH I'm not sure that we can redeem this situation. If two VMs are already running with the same firmware UUID, I think one has to be re-created. In any case, I think it shouldn't be in the scope of this proposal which only aims to prevent this situation from happening in the future.

SGTM, thanks.

   - May necessitate warning mechanisms or tooling for transition.

## API Examples
**Example uuid persistence within VM Status**

```yaml
apiVersion: kubevirt.io/v1alpha3
kind: VirtualMachine
metadata:
  name: mytestvm
status:
   conditions:
      - lastProbeTime: "2024-11-06T01:12:29Z"
        lastTransitionTime: "2024-11-06T01:12:29Z"
        message: Guest VM is not reported as running
        reason: GuestNotRunning
        status: "False"
        type: Ready
   created: true
   runStrategy: Once
   firmwareUUID: "123e4567-e89b-12d3-a456-426614174000"
```

or persist via status condition

```yaml
apiVersion: kubevirt.io/v1alpha3
kind: VirtualMachine
metadata:
  name: mytestvm
status:
   conditions:
      - type: FirmwareUUIDPersisted
        status: "123e4567-e89b-12d3-a456-426614174000"
        reason: UUIDGenerated
        message: "Firmware UUID has been generated and persisted"
        lastTransitionTime: "2024-11-06T01:12:29Z"
```

## Scalability
The proposed changes have no anticipated impact on scalability capabilities of the KubeVirt framework

## Update/Rollback Compatibility
should not affect updates / rollbacks.

## Functional Testing Approach

> Verify that a newly created VMIs has a unique firmware UUID assigned and that this UUID persists across VMI restarts.

Reviewer 4: ```suggestion
Verify that a newly created VMI has a unique firmware UUID assigned and that this UUID persists across VMI restarts.
```

Author: fixed

   - Avoids adding a new API field, fully backward compatible.

   **Cons:**
   - This approach "abuses" the spec by storing data that does not represent the user's desired state.
   - Could lead to cluttered and harder-to-read VM definitions.

2. **Generate UUID via Webhook and Store in the Spec Field**  
   **Description:** If no UUID is specified by the user, a webhook generates a random UUID and assigns it to the VM’s spec field.

   **Pros:**
   - Similar to the first option (no new API field, backward compatible).

   **Cons:**
   - Adds complexity by involving webhooks, which could slow down performance if numerous VMs start simultaneously.
   - May potentially bottleneck `virt-api`.

3. **Add a Firmware UUID Field to VM Status**  
   **Description:** Store the generated UUID within the VM’s status.

   **Pros:**
   - Keeps the spec clean.
   - Focuses on VM status for generated information.

   **Cons:**
   - Could introduce a new API field, affecting long-term maintenance and adding to potential API bloat.
   - The status field is dynamic and frequently updated by controllers, which could make the stored UUID fragile or prone to accidental modification.

> 

Reviewer 2: 5.
* Just before kubevirt is upgraded, persist Firmware.UUID of existing VMs in vm.spec. From that point on, a VM without Firmware.UUID is new.
* From that point on, use vm.metadata.uid if vm.spec.template.spec.Firmware.UUID is not defined.

Pros:
* The upgrade-specific code can be deleted after two or three releases
* Very simple logic post-upgrade
* Disturbance to git-ops is limited to pre-existing (somewhat) buggy VMs.

Cons:
* We need to add a piece of code (to virt-operator?) to implement the upgrade-specific piece.

Reviewer 2: > then we should probably wait for 1-2 releases until all users upgraded their machine, and we will be free to change the generation logic as we pleases

Not sure I follow your comment. I'm suggesting to do a one-off persisting of buggy uuid during KubeVirt upgrade. The new logic (`vm.metadata.uid`) takes effect immediately after that.

Author: understood, my bad

Reviewer 6: @dankenigsberg Patching the spec of a running VM will raise the required restart condition.

* Maintain backward compatibility
* Maintain UUID persistence with VM backup/restore.

## Non Goals

## Definition of Users
End Users: Individuals or organizations running VMs and VMIs on KubeVirt who require consistent firmware UUIDs for their applications.

## User Stories
As an end-user, I expect my VMI to maintain its identity across restarts.

## Repos
Kubevirt/kubevirt

# Design

This section provides a range of alternative solutions for implementing a persistent, universally unique firmware UUID for VMIs.

### 1. Persist UUID in the VM's Spec Field

**Description:**

The firmware UUID is generated when the VMI is created and stored in `vm.spec.template.spec.firmware.uuid`.

To prevent disruptions caused by introducing a new generation logic for firmware UUIDs, we propose a multiphase approach:
* Continue with the current firmware UUID calculation, but store the UUID in spec.
* use alerts / VM conditions in order to notify the users to restart VMs that do not have firmware UUID in their spec.
* After a while, let the community know (i.e. via a mail to mailing list) that in the next version firmware UUID must reside in spec.

> * Switch to using vm.metadata.uid as the new firmware UUID (unless a UUID is already provided in the VM's spec, therefore keeping backward compatibility).

Reviewer 6: 1. I generally dislike patching VM's spec post admission.  This disguises the user intent. 
2. If we are going to keep the code that performs the legacy calculation, we will never get rid of this code.
If we've already patched the spec, use what's written in the spec.

> * Switch to using vm.metadata.uid as the new firmware UUID (unless a UUID is already provided in the VM's spec, therefore keeping backward compatibility).

Author: I don't think it's possible to completely remove the legacy calculation just yet, how the controller will handle old VMs that don't have a VMI to read it's FW ID? 
he has to use the legacy calculation to preserve the backwards compatibility

- Requires additional communication and support, as users may need to update or reconfigure their VMs to prevent disruptions.
- May necessitate tooling to facilitate UUID preservation for compatibility with existing workloads.

---

### 4. Upgrade-Specific Persistence of Firmware UUID

**Description:** Before upgrading KubeVirt, persist the `Firmware.UUID` of existing VMsin `vm.spec`.
After the upgrade, any VM without `Firmware.UUID` is considered new.
For these new VMs, use `vm.metadata.uid` as the firmware UUID if `vm.spec.template.spec.Firmware.UUID` is not defined.

**Pros:**
- Upgrade-specific code can be removed after two or three releases.
- Maintains a straightforward logic post-upgrade, with minimal changes required.
- Limits disturbance to GitOps workflows to pre-existing VMs that had potentially buggy UUID behavior.

**Cons:**
- Requires additional code (likely in `virt-operator`) to handle the upgrade-specific persistence.
- Code should be added to the computation of the "restart required" condition, so that it is not raised if vm.Template.Firmware.UUID equals vmi.Firmware.UUID.

---

> ## What Happens to the Firmware UUID During a Restore?

Reviewer 6: It can be cloned, which will remove the VM specific data, but not restored twice.

* The firmware UUID is part of the VM's spec, or dynamically generated by the VM controller if it is not defined.
* Snapshots created before implementing the firmware UUID persistence mechanism will not include this field.
* As a result, during a restore, a new UUID may be generated if the vm.spec.template.spec.firmware.uuid field is absent.

**Note:** Focusing on persistence ensures that existing behavior remains unchanged, avoiding restore issues while laying a foundation for future UUID management enhancements.

## API Examples
**Example uuid persistence within VM Status**

```yaml
apiVersion: kubevirt.io/v1alpha3
kind: VirtualMachine
metadata:
  name: mytestvm
status:
   conditions:
      - lastProbeTime: "2024-11-06T01:12:29Z"
        lastTransitionTime: "2024-11-06T01:12:29Z"
        message: Guest VM is not reported as running
        reason: GuestNotRunning
        status: "False"
        type: Ready
   created: true
   runStrategy: Once
   firmwareUUID: "123e4567-e89b-12d3-a456-426614174000"
```

## Scalability
The proposed changes have no anticipated impact on scalability capabilities of the KubeVirt framework

## Update/Rollback Compatibility

> should not affect updates / rollbacks.

Reviewer 7: This is where the content of `## What Happens to the Firmware UUID During a Restore?` should land.
Or you should remove this part.

## Functional Testing Approach
Verify that a newly created VMI has a unique firmware UUID assigned and that this UUID persists across VMI restarts.
Validate that the selected persistence mechanism (spec field, status field) stores the UUID.
Include unit tests for any logic introduced in controllers to ensure the UUID is generated and persisted correctly.

# Implementation Phases
- Based on the selected design, either introduce a new field or utilize an existing one 
(such as in spec, status) to store the firmware UUID.
- Update controller logic to check for and persist the UUID, ensuring it is generated only once per VM.

> - Testing: add unit and functional tests

Reviewer 7: I do not think these are valid phases.

You cannot add any logic without unit tests and possibly basic e2e tests.
And you cannot add the API change without utilizing it, not even as a separate PR.

Usually phasing means that  you provide a basic functionality that works, then extend it in the next phases. It can also include feature lifecycle stages, i.e. alpha, beta and GA planning.
BTW, you should mention explicitly that there is no plan to protect this new logic with a FG (the fact it was not mentioned, implied to me that this is the intention).

* Maintain UUID persistence with VM backup/restore.
* Change UUID of currently existing VMs

## Definition of Users
VM owners: who require consistent firmware UUIDs for their applications.
cluster-admin: to ensure VMs have universally unique firmware IDs

## User Stories

### Supported Cases

#### 1. Creating and Starting a New VM
**As a user**,  
I want to create and start a new VM in KubeVirt,  
**so that** it gets a unique, persistent firmware UUID for consistency across reboots, migrations, and clusters.

- **Scenario**: Creating VM `X` in namespace `Y` on Cluster `A` and `B` results in different UUIDs.
- **Expected**: UUIDs are unique universally to avoid conflicts.

---

#### 2. Starting a VM Created by an Older KubeVirt Version
**As a user**,  
I want to start a VM from an older KubeVirt version,  
**so that** it is assigned a persistent UUID if one wasn’t set before.

- **Scenario**: A VM without a UUID starts on an upgraded KubeVirt version.
- **Expected**: The old UUID is persisted in the VM spec to ensure consistency.

## Repos
Kubevirt/kubevirt

## Proposed Solution

### Description
The firmware UUID persistence mechanism will operate as follows:

1. **New VMs**:

>     - If the firmware UUID is not explicitly defined in `vm.spec.template.spec.firmware.uuid`, the mutator webhook will automatically set the firmware UUID to the value of `vm.metadata.uid`.

Reviewer 8: The fact that `vm.metadata.uid` is not set in a CREATE webhook makes this scheme a little clunky. Are we sure that `vm.spec.template.spec.firmware.uuid` will not be used before the VM is updated for the first time? Probably. But just pointing it out.

Also, our more security conscious users may prefer that the value of `vm.metadata.uid` is not used directly, rather hashed with some const. Why leak such information to the guest?

Reviewer 8: Actually, how will you determine new vs existing VM without the ability to know uid `CREATE`?

Author: I adjusted the design, we can generate random UUID in the mutator that is not based on the `vm.metadata.uid`

Reviewer: > I adjusted the design, we can generate random UUID in the mutator that is not based on the `vm.metadata.uid`

How would we ensure that it is universally unique?

Author: cant we use what k8s uses: 

```
package uuid

import (
	"github.com/google/uuid"

	"k8s.io/apimachinery/pkg/types"
)

func NewUUID() types.UID {
	return types.UID(uuid.New().String())
}
```

Author: done

> 

Reviewer 8: What about standalone VMIs? I assume uid will be used as well? The lack of uid in CREATE may be a bigger issue here

Author: could you elaborate please what do you mean the lack of uid in CREATE ? 
in the design when a VM is being created we make sure the firmware uuid is populated by the mutator web-hook if the user didn't provide his own.

Reviewer 8: Webhooks have an operation associated to them (CREATE, UPDATE, DELETE, or CONNECT). The CREATE webhook is called before `metadata.uid` is set. That field is set when the object is persisted to etcd (after webhooks are called). So your scheme of assigning firmware id to `metadata.uid` cannot work for CREATE. 

You will have `metadata.uid` for other operations but if UPDATE is called for a VM with no fw id, how do you know if it is a "new" VM (gets metadata uid) or old VM (hash of name)?

Reviewer: Thank you @mhenriks! That's a good point.
One possible resolution to this is to handle the firmware UUID assignment from the controller instead of the webhooks.

FYI @vladikr

Author: Thank you for clarifying this point, @mhenriks. I now see your concern more clearly. This limitation indeed renders the webhook logic redundant, as it cannot reliably distinguish between new and existing VMs during the create operation due to the absence of the vm.meta.uid at this stage.

Implementing additional logic in the webhook to determine whether a VM is new feels out of scope for a mutation webhook's responsibilities. Instead, this kind of logic would be better suited within the controller, where we can manage such scenarios more effectively.

Author: Also if we move the logic to the controller, and make sure we establish a way to distinguish between new VMs and existing VMs, we won't need to interfere with the restore controller (to try to patch the old legacy calculation for the firmware uuid, the controller will do that for us since the mutation web-hook is out the the scope). 
it's 2 birds with one stone. 

the only con is that the establishment of new VMs vs existing VMs in the controller might be a little bit complex

Reviewer 6: Yes, @mhenriks thank you for steering this in the right direction.
@xpivarc FYI

I'm worried about the controller-based solution since it will add latency when setting this field. There could be another controller that makes assumptions about this field and it may potentially be in a race with the VM controller. I prefer not to update the VM spec fields via a controller.

Perhaps we could just set a random UID?
Alternatively, we could fold back to the idea of storing the UID in the VM status.  

What do you think?

Reviewer 8: Yeah I think assigning a new random UUID is the best we can do for new VMs (CREATE mutating webhook would work). I'm still unclear how how existing VMs can best be be assigned the old style FW id

- **Scenario**: Creating VM `X` in namespace `Y` on Cluster `A` and `B` results in different UUIDs.
- **Expected**: UUIDs are unique universally to avoid conflicts.

---

#### 2. Starting a VM Created by an Older KubeVirt Version
**As a user**,  
I want to start a VM from an older KubeVirt version,  
**so that** it is assigned a persistent UUID if one wasn’t set before.

- **Scenario**: A VM without a UUID starts on an upgraded KubeVirt version.
- **Expected**: The old UUID is persisted in the VM spec to ensure consistency.

## Repos
Kubevirt/kubevirt

## Proposed Solution

### Description
The firmware UUID persistence mechanism will operate as follows:

1. **New VMs**:
    - If the firmware UUID is not explicitly defined in `vm.spec.template.spec.firmware.uuid`, the mutator webhook will automatically set the firmware UUID on create operation.
    - We can use what k8s uses to generate uuids via `"github.com/google/uuid"` package.

2. **Existing VMs**:
    - For VMs created before the upgrade, the VM controller will patch the `vm.spec.template.spec.firmware.uuid` field,
   with a value calculated using the legacy logic (based on the VM name).

>    This ensures backward compatibility.

Reviewer 9: What if we have an annotation that stores the legacy uuid and then patch the old VMs with the new uuid on upgrade? What are the blockers in terms of backward-compatibility?

Reviewer 10: @dasionov After the upgrade, how would you know if a certain VM is from older versions or from current version?

Reviewer 10: @acardace Good point, I think it will work only for running VMs. Those that are off, we probably won't know whether to apply legacy calculations on them.

Author: Every VM that does not have a persisted UUID in the spec should be considered old, as new VMs will receive a random UUID from the mutator webhook. Therefore, it does not matter whether the VMs were running or stopped before the upgrade.

> What if we have an annotation that stores the legacy uuid and then patch the old VMs with the new uuid on upgrade

I Have a vague memory that @EdDev was little bit against annotation solution for this specific bug.

* The non-persistent nature of UUIDs may trigger redundant cloud-init setup upon VM restore. (issue)
* The non-persistent nature of UUIDs may trigger redundant license activation events during VM lifetime i.e. Windows key activation re-occur when power off and power on the VM.

## Goals
* Improve the uniqueness of the VMI firmware UUID to become independent of the VM name and namespace.
* Maintain the UUID persistence over the VM lifecycle.
* Maintain backward compatibility
* Maintain UUID persistence with VM backup/restore.

## Non Goals

## Definition of Users
VM owners: who require consistent firmware UUIDs for their applications.

## User Stories
### Supported cases:
* Creating and Starting a New VM shoul
* starting and old VM that was not running
* restoring an old VM
* restoring a new VM

## Repos
Kubevirt/kubevirt

# Design

This section provides a range of alternative solutions for implementing a persistent, universally unique firmware UUID for VMIs.
Option 1 is preferable as it avoids new API fields, ensures backward compatibility, and has minimal drawbacks.

### 1. Persist UUID in the VM's Spec Field

**Description:**

> if the FW ID is not specified in `vm.spec.template.spec.firmware.uuid` it shell be generated and stored in `vm.spec.template.spec.firmware.uuid`.

Reviewer 2: ```suggestion
if the FW ID is not specified in `vm.spec.template.spec.firmware.uuid` it shall be generated and stored in `vm.spec.template.spec.firmware.uuid`.
```

**Pros:**
- Retains backward compatibility and avoids introducing new API fields.

**Cons:**
- Same concerns as above about the spec field.
- Using webhooks to assign defaults can degrade performance, especially with high volumes of VMs, as `virt-api` may become a bottleneck.
- May compromise scalability, as generating a UUID via webhook adds overhead when multiple VMs start simultaneously.

---

### 3. Create a New Firmware UUID Field in VM Status

**Description:** A new field is added under VM status to store the generated firmware UUID upon first VM boot.

**Pros:**
- Keeps the spec clean, avoiding unnecessary fields in the user-defined configuration.
- Provides a clear separation between user configuration (spec) and system-generated information (status).

**Cons:**
- Introduces a new API field, which could increase API surface area and requires long-term support.
- New fields may impact system performance and resource load.
- Adds to API complexity and may reduce readability.

*Note:* To minimize this con, consider setting the UUID as a condition under VM status instead of introducing a separate field.

---

### 4. Introduce a Breaking Change to Ensure Universal Uniqueness

> **Description:** Modify UUID generation to use a combination of VM name and namespace, ensuring unique UUIDs within the cluster.

Reviewer: > what about universal uniqueness? How would it be ensured?

Using both name + namespace as a hash will ensure uniqueness AFAICT

> **Description:** Modify UUID generation to use a combination of VM name and namespace, ensuring unique UUIDs within the cluster.

Reviewer 2: not universal uniqueness. a different vm on a different cluster would obtain the same uuid if it has the same name and namespace.

> **Description:** Modify UUID generation to use a combination of VM name and namespace, ensuring unique UUIDs within the cluster.

Reviewer: Ah, got you.
We can always add something "random" to the hash, like creation timestamp.

> **Description:** Modify UUID generation to use a combination of VM name and namespace, ensuring unique UUIDs within the cluster.

Author: Would it be possible to clarify the impact of VMs in different clusters having the same UUID if they share the same name and namespace?
Are there specific cases where cross-cluster uniqueness is essential for the UUID to function correctly?

> **Description:** Modify UUID generation to use a combination of VM name and namespace, ensuring unique UUIDs within the cluster.

Reviewer: I guess that @dankenigsberg is referring migrating a VM from a different cluster, e.g. via importing/exporting disks

> **Description:** Modify UUID generation to use a combination of VM name and namespace, ensuring unique UUIDs within the cluster.

Reviewer 2: > Are there specific cases where cross-cluster uniqueness is essential for the UUID to function correctly?

By definition, [UUID](https://en.wikipedia.org/wiki/Universally_unique_identifier)  has to be universally-unique. A lot of services would break if two different VMs present matching identifiers.

> **Description:** Modify UUID generation to use a combination of VM name and namespace, ensuring unique UUIDs within the cluster.

Reviewer 6: Just use the pod medata.UID?
https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#uids

> **Description:** Modify UUID generation to use a combination of VM name and namespace, ensuring unique UUIDs within the cluster.

Author: you mean the vm.meta.uid?

By storing the firmware UUID, we ensure that it remains consistent across VMI restarts.
which is crucial for applications and services that rely on the UUID for identification or licensing purposes.

## Motivation
By definition, UUID must be a universally-unique identifier.
The current implementation for the automatically-computed Firmware UUID is not universally unique: 
Two VMs with the same name get the same UUID, even if they reside in different namespaces or different clusters.
This proposal discusses how to fix this, while ensuring that existing VMs keep their current ID.

## Goals
* Generate and persist universally-unique Firmware.UUID for new VMs
* Keep the Firmware.UUID of pre-existing VMs

## Non Goals

## Definition of Users
End Users: Individuals or organizations running VMs and VMIs on KubeVirt who require consistent firmware UUIDs for their applications.

## User Stories
As an end-user, I expect my VMI to maintain its identity across restarts.

## Repos
Kubevirt/kubevirt

# Design

The introduction of a persistent, universally unique firmware UUID will proceed in two phases:

> ### Phase 1: Preserve Current Generation Method and Persist UUID

Reviewer 2: Not all alternatives need to be broken into phases, so let us not force phases on all designs. "Upgrade-Specific Persistence of Firmware.UUID" is about introduction of upgrade+bugfix in one release.

## Single-Phase Solution for Persistent, Universally Unique UUIDs

### 1. Upgrade-Specific Persistence of Firmware.UUID

**Description:** Prior to upgrading KubeVirt, persist the `Firmware.UUID` of existing VMs in `vm.spec`. 
After the upgrade, any VM without a `Firmware.UUID` in its spec will be considered new.
For these new VMs, the system defaults to using `vm.metadata.uid` as the firmware UUID if `vm.spec.template.spec.Firmware.UUID`is not defined.

**Pros:**
- The upgrade-specific code can be safely removed after two or three releases.
- Simple logic post-upgrade, with a straightforward UUID assignment.

> - Limited disturbance to GitOps workflows, affecting only pre-existing VMs with the current method.

Reviewer 11: This is only true if gitops is setting metadata.uid, correct?

Reviewer 2: @jangel97 what is the benefit of storing this in a new annotation, instead of an already-existing API?

**Cons:**
- Requires additional code (likely in `virt-operator`) to handle the upgrade-specific persistence.

> - Patching the spec of a running VM may trigger a "restart required" condition.

Reviewer 2: Can this be easily avoided in code, if we find that vm's Firmware.UUID has NOT differed from the vmi's.

Reviewer 6: Generally, patching object specs after the admission is not a good practice that hides the user intent.

Reviewer 2: > Generally, patching object specs after the admission is not a good practice that hides the user intent.

Ack. And I'd say that patching during admission is just as bad in this respect.
However in this case, the user never intended us to have a buggy UUID, but must have it persisted. So I don't see here any issue, beside the (limited) disturbance to git-ops.

> Can this be easily avoided in code, if we find that vm's Firmware.UUID has NOT differed from the vmi's?

@dasionov would you refer to this? I believe that the Con should say

`- Code should be added to the computation of the "restart required" condition so that it is not raised if vm.Template.Firmware.UUID equals vmi.Firmware.UUID.`

Author: sry for the delay, fixed

This phase focuses on ensuring VMs retain their UUID across restarts without modifying the existing generation approach.
Below are potential solutions for this phase:

1. **Persist UUID in the VM’s Spec Field**  
   **Description:** The UUID would be generated on the VM’s first boot and stored in the VM’s spec field.

   **Pros:**
   - Straightforward implementation.
   - Avoids adding a new API field, maintaining backward compatibility.

   **Cons:**
   - This approach "abuses" the spec by storing data that does not represent the user's desired state.
   - Could lead to cluttered and harder-to-read VM definitions.

2. **Generate UUID via Webhook and Store in the Spec Field**  
   **Description:** If no UUID is specified by the user, a webhook generates a random UUID and assigns it to the VM’s spec field.

   **Pros:**
   - Similar to the first option, avoiding a new API field and maintaining backward compatibility.

   **Cons:**
   - Adds complexity by involving webhooks, which could slow down performance if many VMs are started simultaneously.
   - May potentially bottleneck `virt-api`.

3. **Add a Firmware UUID Field to VM Status**  

>    **Description:** Store the generated UUID within the VM’s status.

Reviewer 11: Status is only live, it must not be used to store information.

aka any informatoin in the status in only relevant and "persisted" on the runtime platform.
Object smust not be created with a populated status (only empty status is what we want: `status: {}`)

Reviewer: > Object smust not be created with a populated status (only empty status is what we want: `status: {}`)

A VM object won't be created with a populated status, but a VM that was created, booted, then shut off would have a non-empty status.

Am I missing anything?

Reviewer: ping @fabiand :)
just want to understand if I get you correctly.

In the second phase, the approach for generating the firmware UUID will be modified to ensure global uniqueness.
This transition could include changing the UUID generation to consider both the VM name and namespace,
or adopting a pseudo-random hash to guarantee uniqueness across clusters. The solutions for this phase are outlined below:

1. **Use Namespace + Name for UUID Generation**  
   **Description:** Generate the UUID by combining the VM name and namespace, ensuring uniqueness within the cluster.

   **Pros:**
   - Simple and straightforward to implement.
   - Avoids reliance on an external generator for randomness.

>    **Cons:**

Reviewer 2: This is not acceptable, as it is not universally unique. If you want to consider it, please state so as a "con".

Author: fixed

3. **Use `vm.metadata.uid` as the Firmware UUID**  
   **Description:** Use the `vm.metadata.uid`, a unique identifier assigned to each VM by Kubernetes, as the firmware UUID.

   **Pros:**
   - Ensures uniqueness within and across clusters, as `metadata.uid` is universally unique for each VM.
   - Avoids adding new fields or complex generation logic.
   - Simplifies UUID management by leveraging an existing unique identifier.

   **Cons:**

By addressing each phase separately, we can balance immediate workload stability with a gradual transition to globally unique UUIDs. This approach allows for flexibility in implementation and can minimize disruption for end-users.

## API Examples
**Example uuid persistence within VM Status**

```yaml
apiVersion: kubevirt.io/v1alpha3
kind: VirtualMachine
metadata:
  name: mytestvm
status:
   conditions:
      - lastProbeTime: "2024-11-06T01:12:29Z"
        lastTransitionTime: "2024-11-06T01:12:29Z"
        message: Guest VM is not reported as running
        reason: GuestNotRunning
        status: "False"
        type: Ready
   created: true
   runStrategy: Once
   firmwareUUID: "123e4567-e89b-12d3-a456-426614174000"
```

> or persist via status condition

Reviewer 6: I'd argue against a condition. Conditions are generally needed to manage the lifecycle or to be consumed by other controllers.

Reviewer 11: conditions are not for persistence.

>    firmwareUUID: "123e4567-e89b-12d3-a456-426614174000"

Reviewer 11: .status ins only for erporting, not for persitence.

Reviewer: Can you elaborate?
How is spec more persistent than status? Aren't they both persisted in etcd?

Reviewer 6: > .status ins only for erporting, not for persitence.

@fabiand I don't know where this is coming from. Numerous objects across k8s use the status exactly for that purpose, including objects in KubeVirt.

Reviewer 6: The VM object, until it is deleted would have a status - it simply stores the current state of the VM, while the spec holds the desired.

Reviewer 7: Conceptually, it is not what status is meant to show.

Status is meant to represent the current state. However, the FW UUID cannot be "read" from the current state of the VM, because it is stopped.
There is a name to persisting such information on the status: caching.
Status is not about caching, it is about representing the actual. But here the suggestion is to store the actual so it will be the input for a future actual (which comes in the opposite direction).

We actually had this implemented in the network interfaces status and it caused a mess: To calculate the new status on a reconcile cycle, the previous data was considered, potentially dragging incorrect state over cycles indefinitely. We have since dropped this behavior, calculating the interfaces status without depending on the previous state.
But this is still tricky business even today, because multiple components attempt to touch the same list.

Bottom line, while this could be a good solution, caching the data on the status can potentially bite back. There is logic that uses this status information to set the UUID on the next VM instantiation and an opposite logic that reads the VM instantiation spec (or status?) to reflect that actual back to the same status field. Sounds like a loop to me.

Reviewer 6: Even if the VM is stopped the status shows the `current` state at this UUID is stored on the VM boot disk
As long as the VM exists this is its current state, regardless of its phase.

- Adds to API complexity and may reduce readability.

*Note:* To minimize this con, consider setting the UUID as a condition under VM status instead of introducing a separate field.

---

### 4. Introduce a Breaking Change to Ensure Universal Uniqueness

**Description:** Modify UUID generation to a combination of `VM.Name`, `vm.Namespace`, and a random hash
to ensure that UUIDs are universally unique, even across clusters.

Users would be notified of this change in advance through logs or alerts,
and optional tooling could be provided to help users preserve their current firmware UUIDs by adding them to the spec.

**Pros:**
- Ensures universally unique UUIDs, meeting the UUID standard and avoiding conflicts
  when migrating VMs across clusters.
- Keeps the spec clean without repurposing fields.
- Simple to implement without introducing new API fields.

**Cons:**
- This approach is a breaking change, requiring user intervention to preserve existing UUIDs.
- Requires additional communication and support, as users may need to update
  or reconfigure their VMs to prevent disruptions.
- May necessitate tooling to facilitate UUID preservation for compatibility with existing workloads.

> ### 5. Upgrade-Specific Persistence of Firmware UUID

Reviewer 4: I don’t necessarily object to this approach—in fact, it’s likely unavoidable to prevent disruption to existing workloads—but the process for implementing it should be more thoroughly detailed IMO.

Correct me if I'm wrong, but if VMs are stopped during the upgrade, they cannot rely on VMI objects to derive and persist the firmware UUID, as those objects won’t exist. As a result, existing VMs without a defined `spec.firmware.uuid` field would need to populate the persistent UUID field (whatever that may be) using the current method (based on the name). The new method should then be applied exclusively to new VMs.

While this doesn’t cover _all_ use cases, it’s reasonable to assume it addresses the majority—except for scenarios involving (for example) backup and restore, which present a separate set of challenges.

Reviewer 2: I don't understand the reservation that you are raising. Option 5 does not distinguish between running and non-running VMs. During upgrade, they are all changed to persist the buggy uuid. After upgrade, all VMs receive a fresh uuid.

The problem of restoring old VMs that did not persist their buggy uuid sounds important, but not specific to this option. @dasionov can you address it? in the design? I don't have any good idea...

Reviewer 4: IIRC the original idea was to persist the UUID in existing VMs by obtaining it from the VMIs, therefore my comment. Perhaps I mixed up the proposal with implementation, my bad, you can disregard the comment in that case.

**Description:** Before upgrading KubeVirt, persist the `Firmware.UUID` of existing VMsin `vm.spec`.
After the upgrade, any VM without `Firmware.UUID` is considered new.

> For these new VMs, use `vm.metadata.uid` as the firmware UUID if `vm.spec.template.spec.Firmware.UUID` is not defined.

Reviewer: Another problem with this approach is upgrades that skip versions.
IOW, a user can upgrade Kubevirt v1.0 straight to v1.3, skipping v1.1 and v1.2. With this approach, such users will break their VMs.

## Testing

### Existing Tests
- Verify that a newly created VMI has a unique firmware UUID assigned.
- Ensure the UUID persists across VMI restarts.

### Scenarios Needing Coverage

1. **Old VM Patching**:  
   Validate that old VMs without a firmware UUID receive one calculated using the legacy logic.

2. **New VM Creation**:  
   Verify that the firmware UUID is set to via `"github.com/google/uuid"` when not explicitly defined.

# Implementation Phases

1. **Alpha Phase**
    - Persist firmware UUID in the `VM` template spec field.
    - Implement logic for:
        - **New VMs**: Auto-generate and store random UUID if not explicitly provided.
        - **Existing VMs**: Patch the `vm.spec.template.spec.firmware.uuid` field using the legacy calculation.
        - **Restored VMs**: uuid will be added to the restored VMs that don't have desired uuid
    - Cover implementation with unit tests.
    - add e2e test for uuid persistence for restored vms

2. **Beta Phase**
    - Ensure seamless upgrade compatibility for pre-existing VMs.
    - Communicate changes through release notes and detailed documentation.

3. **GA Phase**

>     - in two versions we can assume that all clusters are already upgraded and the old code can be dropped.

Reviewer 8: How will empty UIDs be handled when old code is dropped?

Author: I don’t see a scenario where a VM resource would be processed after the old code has been removed while still having the empty UUID in its spec.

Reviewer 8: It is still an editable field, is it not? I see no mention of making the field immutable after it's set.

It may not be supported, but does the kubevirt operator keep people from updating from back 3+ versions? Even if it does, I know of users that don't use the operator.

_Something_ is going to happen we encounter an empty id and I am curious if we know what. And if not are we planning for this case

Author: >It is still an editable field, is it not? I see no mention of making the field immutable after it's set.

>It may not be supported, but does the kubevirt operator keep people from updating from back 3+ versions? Even if it does, I know of users that don't use the operator.

>Something is going to happen we encounter an empty id and I am curious if we know what. And if not are we planning for this case

Thank you for the detailed explanation—you raise a valid point once again.
We should definitely discuss this further before finalizing the proposal.

IMO, if a user removes the field for any reason, leaving it empty, it makes sense for the controller to generate a new random UUID, similar to how the mutator behaves on creation.

# Overview
This proposal introduces a mechanism to persist the firmware UUID of a Virtual Machine Instance (VMI) in KubeVirt.
By storing the firmware UUID, we ensure that it remains consistent across VMI restarts.
which is crucial for applications and services that rely on the UUID for identification or licensing purposes.

## Motivation
By definition, UUID must be a universally-unique identifier.
The current implementation for the automatically-computed Firmware UUID is not universally unique: 
Two VMs with the same name get the same UUID, even if they reside in different namespaces or different clusters.

> This proposal discusses how to fix this, while ensuring that existing VMs keep their current ID.

Reviewer 10: I would re-phrase it as follows:
I would extend the content of this section with the following user stories:
* Duplicate UUIDs cause burden with third-party automation frameworks such as Gitops tools, making it difficult to maintain unique VM UUID. Such tools have to manually assign unique UUIDs to VMs, thus also manage them externally.
* The non-persistent nature of UUIDs may trigger redundant license activation events during VM lifetime i.e. Windows key activation re-occur when power off and power on the VM.

Author: fixed

> ## Goals
> * Generate and persist universally-unique Firmware.UUID for new VMs
> * Keep the Firmware.UUID of pre-existing VMs

Reviewer 10: ```suggestion
## Goals
* Improve the uniqueness of the VM UUID to become independent of the VM name and namespace.
* Maintain the UUID persistence over the VM lifecycle.
* Maintain backward compatibility
* Maintain UUID persistence with VM backup/restore.
```

Author: done

* The non-persistent nature of UUIDs may trigger redundant license activation events during VM lifetime i.e. Windows key activation re-occur when power off and power on the VM.

## Goals
* Improve the uniqueness of the VMI firmware UUID to become independent of the VM name and namespace.
* Maintain the UUID persistence over the VM lifecycle.
* Maintain backward compatibility
* Maintain UUID persistence with VM backup/restore.

## Non Goals

## Definition of Users
End Users: Individuals or organizations running VMs and VMIs on KubeVirt who require consistent firmware UUIDs for their applications.

## User Stories
As an end-user, I expect my VMI to maintain its identity across restarts.

## Repos
Kubevirt/kubevirt

# Design

This section provides a range of alternative solutions for implementing a persistent,
universally unique firmware UUID for Virtual Machines Instances (VMIs) in KubeVirt.

### 1. Persist UUID in the VM's Spec Field

**Description:** The UUID is generated when the VMI is being created and is then saved to the VM's spec field.

**Pros:**
- Straightforward implementation.
- Avoids the need to introduce new API fields.

> - Fully backward compatible.

Reviewer 6: Wouldn't this break existing VMs? After a shutdown, these will get a new UUID.

Author: I am relying on the assumption that we can capture and store the firmware UUID of existing VMIs before they are shut down, in the first sync of the newer version of virt-controller.  

In the case of the first option, the approach is to store the UUID in `vm.spec.template.spec.firmware.uuid`. This ensures that the UUID is retained and reused across restarts, preventing any disruption to existing VMs.

Reviewer: > After a shutdown, these will get a new UUID.

IIUC the idea here is a multi-phased approach, in which UUID will first be saved to the spec and then (after a few releases) the way of computing the UUID will change. IOW this approach is similar to the status field approach but with a spec field.

**Cons:**
- Potential misuse of the spec field, which ideally should only reflect the user’s intended configuration.
- The majority of users may not consider the UUID part of the desired VM state, making this field irrelevant from their perspective.

> - Can lead to longer and more complex VM definitions, impacting readability and management.

Reviewer 12: How do you ensure backward compatibility here?

Author: To ensure backward compatibility, I would sync the VMI and check the `vm.spec.template.spec.firmware.uuid` field.  

- If the field already contains a UUID, I would reuse it.  
- If it is empty, I would retrieve the existing firmware UUID from the VMI and store it in the VM spec.  

This approach ensures that no new UUIDs are generated for existing VMIs, preserving their original identity.

Reviewer: Please take a look at https://github.com/kubevirt/community/pull/347#discussion_r1848036967

---

### 2. Generate a Default UUID via Webhook and Set it to the Spec Field

**Description:** A webhook generates a UUID if the user does not provide one. This UUID is then set in the VM's spec field.

**Pros:**
- Retains backward compatibility and avoids introducing new API fields.

**Cons:**
- Same concerns as above about the spec field.

> - Using webhooks to assign defaults can degrade performance, especially with high volumes of VMs, as `virt-api` may become a bottleneck.

Reviewer 12: How does this degrade a performance if the webhook already exists? How is this a bottleneck if the API can be scaled?

Author: i guess maybe not everyone would want to scale the entire virt-api component just to get more web-hooks power.. IMHO

- May compromise scalability, as generating a UUID via webhook adds overhead when multiple VMs start simultaneously.

> ---

Reviewer 12: You should describe why is it backward compatible (The controller uses same function to "generate" UUID in case of old vm where old vm is equal to vm that does not have the UUID set)

### 3. Create a New Firmware UUID Field in VM Status

**Description:** A new field is added under VM status to store the generated firmware UUID upon first VM boot.

**Pros:**
- Keeps the spec clean, avoiding unnecessary fields in the user-defined configuration.
- Provides a clear separation between user configuration (spec) and system-generated information (status).

**Cons:**
- Introduces a new API field, which could increase API surface area and requires long-term support.

> - New fields may impact system performance and resource load.

Reviewer 6: how?

> - Adds to API complexity and may reduce readability.

Reviewer 6: I don't understand this con, can you elaborate?

*Note:* To minimize this con, consider setting the UUID as a condition under VM status instead of introducing a separate field.

> 

Reviewer 12: Again how do we ensure the existing VMs retain stable UUID?

Reviewer 6: I think the flow of option 3 is not well explained - at least from what I had in mind.
I see this working the following: 

VM controller: 
   On start:
       if the user explicitly set the `vm.spec.template.spec.Firmware.UUID` - propagate this to `vmi.spec.Firmware.UUID`
       Otherwise (user didn't set firmware UUID explicitly):
            sets the `vmi.spec.Firmware.UUID` to `vm.metadata.uid` if `vm.Status.FirmwareUUID` is empty (because it's a newly created VM - no VMI existed before)
            if vm.status.FirmwareUUID is set - propagate it to `vmi.spec.Firmware.UUID`

    On Sync:
        if `vm.status.FirmwareUUID` is empty set it using `vmi.spec.Firmware.UUID` 

This approach preserves backward compatibility and is still gitops friendly

Reviewer 10: @xpivarc @vladikr I think we can re-use vTPM backend storage approach here, instead of using the VM status as the carrier for this data, WDYT?

Reviewer 10: As I imagine it, by using the same approach as with vTPM backend storage it would be easy to manage the VM Cloning, snapshot and restore aspects, i.e. we would not clone the UUID PVC as it should be unique just like the vTPM

---

> ### 4. Introduce a Breaking Change to Ensure Universal Uniqueness

Reviewer 12: -1

**Description:** Modify UUID generation to a combination of `VM.Name`, `vm.Namespace`, and a random hash
to ensure that UUIDs are universally unique, even across clusters.

Users would be notified of this change in advance through logs or alerts,
and optional tooling could be provided to help users preserve their current firmware UUIDs by adding them to the spec.

**Pros:**
- Ensures universally unique UUIDs, meeting the UUID standard and avoiding conflicts
  when migrating VMs across clusters.
- Keeps the spec clean without repurposing fields.
- Simple to implement without introducing new API fields.

**Cons:**
- This approach is a breaking change, requiring user intervention to preserve existing UUIDs.
- Requires additional communication and support, as users may need to update
  or reconfigure their VMs to prevent disruptions.
- May necessitate tooling to facilitate UUID preservation for compatibility with existing workloads.

> ### 5. Upgrade-Specific Persistence of Firmware UUID

Reviewer 12: This would require having 2 releases to fix this issue or some kind of coordination of the upgrade that is not available today.

in a way that best balances stability, performance, and usability for KubeVirt users.

## API Examples
**Example uuid persistence within VM Status**

```yaml
apiVersion: kubevirt.io/v1alpha3
kind: VirtualMachine
metadata:
  name: mytestvm
status:
   conditions:
      - lastProbeTime: "2024-11-06T01:12:29Z"
        lastTransitionTime: "2024-11-06T01:12:29Z"
        message: Guest VM is not reported as running
        reason: GuestNotRunning
        status: "False"
        type: Ready
   created: true
   runStrategy: Once
   firmwareUUID: "123e4567-e89b-12d3-a456-426614174000"
```

or persist via status condition

```yaml
apiVersion: kubevirt.io/v1alpha3
kind: VirtualMachine
metadata:
  name: mytestvm
status:
   conditions:
      - type: FirmwareUUIDPersisted
        status: "123e4567-e89b-12d3-a456-426614174000"
        reason: UUIDGenerated
        message: "Firmware UUID has been generated and persisted"
        lastTransitionTime: "2024-11-06T01:12:29Z"
```

## Scalability
The proposed changes have no anticipated impact on scalability capabilities of the KubeVirt framework

## Update/Rollback Compatibility
should not affect updates / rollbacks.

## Functional Testing Approach

> Verify that a newly created VMI has a unique firmware UUID assigned and that this UUID persists across VMI restarts.

Reviewer 12: Don't we have this test already?
What we miss is upgrade compatibility.

Author: can we even test that?

Reviewer 7: You could express that some parts are tested as e2e tests and others as unit tests.
You could detail the scenarios which needs coverage.

## Repos
Kubevirt/kubevirt

# Design

This section provides a range of alternative solutions for implementing a persistent,
universally unique firmware UUID for Virtual Machines Instances (VMIs) in KubeVirt.

### 1. Persist UUID in the VM's Spec Field

**Description:**

The firmware UUID is generated when the VMI is created and stored in `vm.spec.template.spec.firmware.uuid`.
This approach ensures backward compatibility because the controller will use the same function to generate a UUID for VMs that do not already have the UUID set.

**Pros:**

- **Simple Implementation:** Requires minimal changes to existing logic.
- **Backward Compatibility:** Preserves existing UUIDs by storing them in the spec.
- **No New API Fields:** Avoids adding new fields to the API.

**Cons:**

- **Spec Misuse:** Spec fields should not store system-generated data.
- **Low Relevance:** Users may not see the UUID as part of the desired state.

---

### 2. Generate a Default UUID via Webhook and Set it to the Spec Field

**Description:** A webhook generates a UUID if the user does not provide one. This UUID is then set in the VM's spec field.

> This approach ensures backward compatibility because the controller will use the same function to generate a UUID for VMs that do not already have the UUID set.

Reviewer 7: I think this got drop in the last update, so the comment is irrelevant now.

| 1      | Existing stopped VM with empty UUID            | VM controller patches VM template spec using legacy logic.                                                                           | Controller checks status field; if missing, uses legacy logic and persists UUID in status.                                 | Controller checks annotation; if missing, uses legacy logic and persists UUID in annotation.                                    |
| 2      | Existing stopped VM with UUID                  | No operation (UUID is already set and remains unchanged).                                                                            | No operation (UUID is already set and remains unchanged).                                                                  | No operation (UUID is already set and remains unchanged).                                                                       |

> | 3      | Existing running VM with empty UUID            | UUID is copied from VMI to VM by the controller using legacy logic.                                                                  | UUID is copied from VMI to VM, then persisted in status field.                                                             | UUID is copied from VMI to VM, then persisted in an annotation.                                                                 |

Reviewer 2: ```suggestion
| 3      | Existing running VM with empty UUID            | UUID is copied from VMI to VM by the controller using legacy logic.                                                                  | UUID is copied from VMI to VM, then persisted in status field.                                                             | UUID is copied from VMI to VM, then persisted in an annotation.                                                                 |
```
For the Proposed Design, running and non-running VMs with no UUID can use the same `VM controller patches VM template spec using legacy logic.` Why involve the VMI?

Author: fixed

- **Pros**:
   - Prevents the bug from spreading to new clusters.
   - Buys time to think of a more robust solution.
- **Cons**:
   - Does not solve the problem for existing clusters.
   - Adds complexity by requiring dual logic for UUID generation.

## Scalability
The proposed changes have no anticipated impact on scalability capabilities of the KubeVirt framework

## Update/Rollback Compatibility
Since this design includes additional logic in the restore controller to preserve UUID compatibility during existing backups, it does not impact updates or rollbacks.

## Testing

### Existing Tests
- Verify that a newly created VMI has a unique firmware UUID assigned.
- Ensure the UUID persists across VMI restarts.

### Scenarios Needing Coverage

1. **Old VM Patching**:  
   Validate that old VMs both Running/Stopped without a firmware UUID receive one calculated using the legacy logic.

2. **New VM Creation**:  
   Verify that the firmware UUID is set to via `"github.com/google/uuid"` when not explicitly defined.

# Implementation Phases

1. **First Phase**
    - Persist firmware UUID in the `VM` template spec field.
    - Implement logic for:

>         - **New VMs**: Auto-generate and store random UUID if not explicitly provided in the mutation-webhook.

Reviewer 2: I think the English is clearer with
```suggestion
        - **New VMs**: Auto-generate and store a random UUID in the mutation-webhook if not explicitly provided.
```

Author: done

# Overview
This proposal introduces a mechanism to persist the firmware UUID of a Virtual Machine Instance (VMI) in KubeVirt. By storing the firmware UUID, we ensure that it remains consistent across VMI restarts, which is crucial for applications and services that rely on the UUID for identification or licensing purposes.

## Motivation

> In the current implementation, the firmware UUID of a Virtual Machine Instance (VMI) is generated dynamically during the creation of the VMI. This means that if the generation logic is changed in the future, the VMI may be assigned a different firmware UUID. This behavior can cause issues for applications that depend on a stable UUID.
> Furthermore, we plan to enhance the firmware UUID generation logic to include the VM's Name and Namespace, ensuring uniqueness across different namespaces. However, changing the generation logic could inadvertently assign new UUIDs to existing VMs upon restart, leading to potential disruptions.
> By persisting the firmware UUID in the VM's status, we can maintain consistency across restarts and future-proof existing VMs against changes in the UUID generation logic.

Reviewer 2: I think that the motivation is simpler and starker. 

```suggestion
By definition, UUID must be a universally-unique identifier. The current implementation for the automatically-computed Firmware UUID is not universally unique: Two VMs with the same name get the same UUID, even if they reside in different namespace or different clusters. This proposal discusses how to fix this, while ensuring that already-identified VMs keep their current ID.
```

Author: fixed

> By persisting the firmware UUID in the VM's status, we can maintain consistency across restarts and future-proof existing VMs against changes in the UUID generation logic.

Reviewer: small nit: when dealing with text / markdown files, it's much easier to review if the lines are short (i.e. 1 line per sentence). The outcome markdown will look the same, but then the reviewer is able to comment with smaller granularity (that is, not select a huge amount of text for a comment).

Author: fixed

## Goals

> - Persist Firmware UUID: Store the firmware UUID within the VM's status to ensure it remains consistent across VMI restarts.

Reviewer 2: keeping it in status is one possible implementation. not a goal.
```suggestion
- Persist Firmware UUID: Store the firmware UUID to ensure it remains consistent across VMI restarts.
```

Author: fixed

- Maintain Backward Compatibility: Ensure that existing VMIs without a persisted firmware UUID continue to operate without issues.

> - Prepare for Future Enhancements: Allow for future changes to the firmware UUID generation logic without impacting existing VMIs.

Reviewer 2: I would state the goals a bit differently:
* Generate and persist universally-unique Firmware.UUID for new VMs
* Keep the Firmware.UUID of pre-existing VMs

Author: done

## Non Goals

> - Changing Firmware UUID Generation Logic Now: This proposal does not include modifying the current firmware UUID generation to incorporate vm.Name and vm.Namespace.

Reviewer 2: Actually, I think that now is the time to design the change. The implementation can wait but we need to think ahead about the desired state.

End Users: Individuals or organizations running VMs and VMIs on KubeVirt who require consistent firmware UUIDs for their applications.

## User Stories
As an end-user, I expect my VMI to maintain its identity across restarts.

## Repos
Kubevirt/kubevirt

# Design
We will introduce a new field FirmwareUUID in the status section of the VirtualMachine resource.
The virt-controller will use this field to persist the firmware UUID (if its empty)

## API Examples
**Example VM Status**

```yaml
apiVersion: kubevirt.io/v1alpha3
kind: VirtualMachine
metadata:
  name: mytestvm
spec:
  runningStrategy: Halted
  template:
    spec:
      architecture: arm64
      domain:
        devices:
          disks:
          - name: disk1
            disk:
              bus: virtio
          interfaces:
          - name: network1
            bridge: {}
      networks:
              - name: network1
                multus:
                  networkName: mytestvm/mytestvm
      volumes:
           - name: disk1
             emptyDisk:
               capacity: 50G
status:
  created: true
  desiredGeneration: 2
  observedGeneration: 2
  printableStatus: ErrImagePull
  runStrategy: Once

>   firmwareUUID: "123e4567-e89b-12d3-a456-426614174000"

Reviewer: Even if we go with this approach we should list a few alternatives for where the new field should reside. Placing it at the top level isn't necessarily the best choice IMO.

An alternative could be to save the UUID as a VM condition:
https://github.com/kubevirt/kubevirt/blob/ced9fe1b5104daa431bed6b65ea690639fe3b79f/staging/src/kubevirt.io/api/core/v1/types.go#L1659-L1660

Conditions are easier to drop in the future, as they do not change the VM's schema.

Author: true
